### PR TITLE
Add `feature.blob_dog` and `feature.blob_log` support for 3D images

### PR DESCRIPTION
**Pull request by emmanuelle**
It was a low-hanging fruit to adapt `feature.blob_dog` and `feature.blob_log` to be 3-D compatible, so I did it. Docstrings are a bit clumsier as a result, suggestions for improvements are welcome.

### CODE REVIEW

**Comment by emmanuelle** [comment]
I can see two possible improvements of these two functions 
- when pruning blobs (40 to 60% of the computing time in the example of the gallery), iterating over all pairs of blobs is quite brute-force. It would be possible to use ``scipy.spatial.cKDTree.query_pairs`` (with a distance given by the maximum radius of a blob?) to decrease the number of pairs. @vighneshbirodkar had you consider this possibility? I could give it a try to see if it reduces indeed the computing time.

- since the different filtered versions of the original image can result in a high memory cost, we could have a kw argument mmap=True or False, and, if True, memory map the cube of data.

**Comment by vighneshbirodkar** [comment]
Hello 
@stefanv had suggested using KDTree but I never got around to trying it. Both your suggestions are excellent, the mmap argument will be specially useful for 3D images. 

Querying with the radius seems like a good idea .

File: skimage/feature/blob.py [hunk 1]
```diff
@@ -193,17 +202,17 @@ def blob_dog(image, min_sigma=1, max_sigma=50, sigma_ratio=1.6, threshold=2.0,
...
-                                  footprint=np.ones((3, 3, 3)),
+                                  footprint=np.ones((3,) * (image.ndim + 1)),
                                   threshold_rel=0.0,
                                   exclude_border=False)
     # Convert local_maxima to float64
     lm = local_maxima.astype(np.float64)
     # Convert the last index to its corresponding scale value
-    lm[:, 2] = sigma_list[local_maxima[:, 2]]
+    lm[:, -1] = sigma_list[local_maxima[:, -1]]
     local_maxima = lm
```
Comment (sciunto) [inline skimage/feature/blob.py]: Do we need this line? Same below.

**Comment by emmanuelle** [comment]
@sciunto not sure I understand your comment. This line replaces indices by the corresponding values of standard deviations, that are used later on in ``_prune_blobs``. So I would say that yes, it is needed.

**Comment by emmanuelle** [comment]
It appears that ``np.stack`` was only introduced in NumPy 1.10, making Travis fail with 1.7. I need to change this.

File: skimage/feature/blob.py (hunk 1 above)
Comment (sciunto) [inline skimage/feature/blob.py]: Can't we return return `_prune_blobs(lm, overlap)`?

Comment (emmanuelle) [inline skimage/feature/blob.py]: OK, I was thinking of the wrong line. Yes, we can remove this. Will do.

Comment (emmanuelle) [inline skimage/feature/blob.py]: @sciunto in French time zone, maybe :-) ?

Comment (sciunto) [inline skimage/feature/blob.py]: back to Paris area tomorrow :)

**Comment by emmanuelle** [comment]
I don't really understand the AppVeyor but I think it has nothing to do with this PR (could someone please check?).

**Comment by sciunto** [comment]
In messages, you can read "Build execution time has reached the maximum allowed time for your plan (60 minutes).". Just too slow :)

**Comment by emmanuelle** [comment]
I implemented the faster version of ``_prune_blobs`` using a cKDTree, and the speed gain is important (for the gallery example, ``blob_log`` runs more than two times faster), although the code is slightly less elegant (I had to remove the nice ``itertools.combination``, too bad!). 

I think this PR is ready for another round of review, since I have made all the required changes.

**Comment by stefanv** [comment]
@emmanuelle Nice work!  Do you have timings before and after?

**Comment by emmanuelle** [comment]
@stefanv See my comment above :-). On my box, ``blob_log`` took 0.8s instead of 2.1.

**Comment by emmanuelle** [comment]
And for a very small number of blobs, filtering + concatenating will take much more time than computing the KDTree.

File: skimage/feature/blob.py [hunk 2]
```diff
@@ -25,26 +26,27 @@ def _blob_overlap(blob1, blob2):
     Parameters
     ----------
     blob1 : sequence
-        A sequence of ``(y,x,sigma)``, where ``x,y`` are coordinates of blob
-        and sigma is the standard deviation of the Gaussian kernel which
-        detected the blob.
+        A sequence of ``(r,c,sigma)`` or ``(f,r,c,sigma)``, where ``r,c``
```
Comment (jni) [inline skimage/feature/blob.py]: I think we went with `p` for `plane`?

... Indeed we [did](http://scikit-image.org/docs/dev/user_guide/numpy_images.html#coordinate-conventions). =)

Comment (jni) [inline skimage/feature/blob.py]: And this is particularly important given your use of `f` in the output. =)

File: skimage/feature/blob.py [hunk 3]
```diff
@@ -25,26 +26,27 @@ def _blob_overlap(blob1, blob2):
...
 
     Returns
     -------
     f : float
-        Fraction of overlapped area.
+        Fraction of overlapped area (or volume in 3D).
     """
-    root2 = sqrt(2)
+    n_dim = len(blob1) - 1
+    root2 = sqrt(n_dim)
```
Comment (jni) [inline skimage/feature/blob.py]: maybe rename this to `root_n`?

File: skimage/feature/blob.py [hunk 4]
```diff
@@ -25,26 +26,27 @@ def _blob_overlap(blob1, blob2):
...
 
     # extent of the blob is given by sqrt(2)*scale
-    r1 = blob1[2] * root2
-    r2 = blob2[2] * root2
+    r1 = blob1[-1] * root2
+    r2 = blob2[-1] * root2
 
-    d = hypot(blob1[0] - blob2[0], blob1[1] - blob2[1])
+    d = sqrt(np.sum((blob1 - blob2)**2))
```
Comment (jni) [inline skimage/feature/blob.py]: Should this not be `blob1[:-1] - blob2[:-1]`?

File: skimage/feature/blob.py [hunk 5]
```diff
@@ -76,9 +85,10 @@ def _prune_blobs(blobs_array, overlap):
     Parameters
     ----------
     blobs_array : ndarray
-        A 2d array with each row representing 3 values, ``(y,x,sigma)``
-        where ``(y,x)`` are coordinates of the blob and ``sigma`` is the
-        standard deviation of the Gaussian kernel which detected the blob.
+        A 2d array with each row representing 3 (or 4) values, ``(r,c,sigma)``
+        or ``(f,r,c,sigma)`` in 3D,  where ``(r,c)`` (``(f,r,c)``) are
```
Comment (jni) [inline skimage/feature/blob.py]: As above, `f` -> `p`

File: skimage/feature/blob.py [hunk 6]
```diff
@@ -25,26 +26,27 @@ def _blob_overlap(blob1, blob2):
     Parameters
     ----------
     blob1 : sequence
-        A sequence of ``(y,x,sigma)``, where ``x,y`` are coordinates of blob
-        and sigma is the standard deviation of the Gaussian kernel which
-        detected the blob.
+        A sequence of ``(r,c,sigma)`` or ``(f,r,c,sigma)``, where ``r,c``
+        (or ``(f,r,c,sigma)``) are coordinates of blob and ``sigma`` is
+        the standard deviation of the Gaussian kernel which detected the blob.
```
Comment (jni) [inline skimage/feature/blob.py]: I would also point out that this needs to be a sequence of arrays, since you do `blob1 - blob2`, which would fail for e.g. tuples. Alternatively, you could add the lines `blob1 = np.asanyarray(blob1)` and `blob2 = np.asanyarray(blob2)`.

File: skimage/feature/blob.py [hunk 7]
```diff
@@ -53,21 +55,28 @@ def _blob_overlap(blob1, blob2):
...
 
-    ratio2 = (d ** 2 + r2 ** 2 - r1 ** 2) / (2 * d * r2)
-    ratio2 = np.clip(ratio2, -1, 1)
-    acos2 = arccos(ratio2)
+        a = -d + r2 + r1
+        b = d - r2 + r1
+        c = d + r2 - r1
+        d = d + r2 + r1
+        area = r1 ** 2 * acos1 + r2 ** 2 * acos2 \
+              - 0.5 * sqrt(abs(a * b * c * d))
```
Comment (jni) [inline skimage/feature/blob.py]: PEP8 suggests parens for line continuations, and linebreaks *after* binary operators (with some further tweaks from me):

```python
area = (acos1 * r1**2 + acos2 * r2**2 -
        0.5 * sqrt(abs(a * b * c * d)))
```

File: skimage/feature/blob.py [hunk 8]
```diff
@@ -405,16 +434,21 @@ def blob_doh(image, min_sigma=1, max_sigma=30, num_sigma=10, threshold=0.01,
...
 
     hessian_images = [_hessian_matrix_det(image, s) for s in sigma_list]
-    image_cube = np.dstack(hessian_images)
+
+    # Replace by image_cube = np.stack(hessian_images, axis=-1)
+    # When we upgrade minimal requirements to NumPy 1.10
+    sl = (slice(None),) * image.ndim + (np.newaxis,)
+    arrays = [np.asanyarray(arr) for arr in hessian_images]
+    extended_arrays = [arr[sl] for arr in arrays]
+    image_cube = np.concatenate(extended_arrays, axis=-1)
```
Comment (jni) [inline skimage/feature/blob.py]: Sheesh, this is repeated quite a few times — maybe put it in its own function?

Comment (jni) [inline skimage/feature/blob.py]: You can even include the call to `peak_local_max`

**Comment by jni** [comment]
@emmanuelle what a great addition! =) I've added some comments, most minor but one quite serious — could you fix it and add a corresponding test? (Or tell me why I'm wrong. =)

File: skimage/feature/blob.py [hunk 9]
```diff
@@ -5,6 +5,7 @@
 import math
 from math import sqrt, hypot, log
 from numpy import arccos
```
Comment (soupault) [inline skimage/feature/blob.py]: Could we get rid of this import and use `np.arccos` instead?

File: skimage/feature/blob.py (hunk 6 above)
Comment (soupault) [inline skimage/feature/blob.py]: I agree. Could we ask here for `ndarray` instead (same as in the functions below)?

File: skimage/feature/blob.py (hunk 8 above)
Comment (vighneshbirodkar) [inline skimage/feature/blob.py]: Why is this needed here in the first place  ? The function supports only 2d blobs

File: skimage/feature/blob.py (hunk 4 above)
Comment (emmanuelle) [inline skimage/feature/blob.py]: oh yes, that's a good catch. Thank you!!!

File: skimage/feature/blob.py (hunk 6 above)
Comment (emmanuelle) [inline skimage/feature/blob.py]: Since this is an internal utility function, I will go for the first solution (change the docstring not the code).

File: skimage/feature/blob.py (hunk 8 above)
Comment (emmanuelle) [inline skimage/feature/blob.py]: I first thought that I could port it to 3D as well :-). But ``_hessian_matrix_det`` doesn't take 3D arguments. OK, I'll revert to the more simple syntax here, for simplicity.

Comment (emmanuelle) [inline skimage/feature/blob.py]: @jni only twice now :-).

File: skimage/feature/blob.py (hunk 9 above)
Comment (emmanuelle) [inline skimage/feature/blob.py]: ok, actually ``math.acos`` might be even better for scalar value.

**Comment by emmanuelle** [comment]
Thank you all to for the review, with a special mention to @jni for detecting a silly bug that I had introduced. There is now a test to check that the output of ``_blob_overlap`` against the analytical formula.

However, when looking for a good test for this issue, I stumbled across (what I consider to be) a bug, which I didn't introduce. Indeed, if you have two overlapping circles, it's almost impossible to have the algorithm returning the two blobs, even with a large value of overlap. Example:
```
>>> img = np.zeros((256, 256))
>>> xs, ys = circle(120, 120, 20)
>>> img[xs, ys] = 1
>>> xs, ys = circle(100, 120, 20)
>>> img[xs, ys] = 1
>>> feature.blob_log(img,  min_sigma=10, max_sigma=30, overlap=0.6)
array([[ 110.        ,  120.        ,   18.88888889]])
```
The actual overlap between the two disks is 0.39, but since the algorithm detects a larger blob centered between the two circles, it keeps only this one (since the overlap between this fake blob and the real circles is much larger than the overlap between the two circles).

@vighneshbirodkar did you run across this kind of problems? Is the algorithm meant to detect only well separated blobs or also blobs that are "glued together"?

**Comment by jni** [comment]
@emmanuelle @vighneshbirodkar I don't know the algorithm but I remember reading about a similar issue with STORM microscopy. The basic idea is for every blob, check whether two smaller blobs would provide a better fit (by some error metric which I assume is built into blob_log), and if so, discard the bigger blob.

I'm not sure what the best way to implement this is, but it doesn't seem impossible.

I also don't know whether this should hold the PR back. Thoughts?

Finally, I'm pretty happy to bump our scipy requirement to 0.10 at least...

**Comment by emmanuelle** [comment]
@jni indeed the issue with overlapping blobs should not hold the PR, since it does not change the behavior of the function. However, it'd be curious to hear from people using this function at the moment, since this issue with overlapping blobs sounds like to a serious limitation (I was planning to use the function on my tomography images, but if overlapping blobs are not correctly detected I cannot really use it). 

About scipy 0.10, it does not have ``cKDTree.query_pairs`` either... 

Could we have a tool that would help us getting rid of unnecessary code when we bump the minimal requirements? It could be something as simple as a text file with pointers to functions, but maybe somebody has a better idea...

**Comment by vighneshbirodkar** [comment]
I don't recall running into this problem, but the method isn't really aware of the shapes of objects, so I can understand why it's doing that.

**Comment by emmanuelle** [comment]
@vighneshbirodkar thank you!

Waiting for final review, then!

**Comment by emmanuelle** [comment]
ping ping, PR ready I think :-)

**Comment by emmanuelle** [comment]
@scikit-image/core it'd be lovely if someone could review this PR! :-)

File: skimage/feature/blob.py [hunk 10]
```diff
@@ -24,50 +23,58 @@ def _blob_overlap(blob1, blob2):
...
-        A sequence of ``(y,x,sigma)``, where ``x,y`` are coordinates of blob
-        and sigma is the standard deviation of the Gaussian kernel which
-        detected the blob.
-    blob2 : sequence
-        A sequence of ``(y,x,sigma)``, where ``x,y`` are coordinates of blob
-        and sigma is the standard deviation of the Gaussian kernel which
-        detected the blob.
+    blob1 : sequence of arrays
+        A sequence of ``(row, col, sigma)`` or ``(pln, row, col, sigma)``,
+        where ``row, col`` (or ``(pln, row, col, sigma)``) are coordinates
```
Comment (jni) [inline skimage/feature/blob.py]: ``(pln, row, col)`` -- sigma is extra in this context. =)

File: skimage/feature/blob.py [hunk 11]
```diff
@@ -24,50 +23,58 @@ def _blob_overlap(blob1, blob2):
...
+        of blob and ``sigma`` is the standard deviation of the Gaussian kernel
+        which detected the blob.
+    blob2 : sequence of arrays
+        A sequence of ``(row, col, sigma)`` or ``(pln, row, col, sigma)``,
+        where ``row, col`` (or ``(pln, row, col, sigma)``) are coordinates
```
Comment (jni) [inline skimage/feature/blob.py]: As above

File: skimage/feature/blob.py [hunk 12]
```diff
@@ -24,50 +23,58 @@ def _blob_overlap(blob1, blob2):
...
-    acos1 = arccos(ratio1)
+    if n_dim == 2:
+        ratio1 = (d ** 2 + r1 ** 2 - r2 ** 2) / (2. * d * r1)
+        ratio1 = np.clip(ratio1, -1, 1)
+        acos1 = math.acos(ratio1)
 
-    ratio2 = (d ** 2 + r2 ** 2 - r1 ** 2) / (2 * d * r2)
-    ratio2 = np.clip(ratio2, -1, 1)
-    acos2 = arccos(ratio2)
+        ratio2 = (d ** 2 + r2 ** 2 - r1 ** 2) / (2. * d * r2)
```
Comment (jni) [inline skimage/feature/blob.py]: @emmanuelle why is the `.` necessary? If it's for Python 2, I would prefer a `from __future__ import division`

File: skimage/feature/blob.py [hunk 13]
```diff
@@ -24,50 +23,58 @@ def _blob_overlap(blob1, blob2):
...
+        d = d + r2 + r1
+        area = (r1 ** 2 * acos1 + r2 ** 2 * acos2 -
+                0.5 * sqrt(abs(a * b * c * d)))
+        return area / (math.pi * (min(r1, r2) ** 2))
 
-    return area / (math.pi * (min(r1, r2) ** 2))
+    else:  # http://mathworld.wolfram.com/Sphere-SphereIntersection.html
+        vol = math.pi / (12 * d) * (r1 + r2 - d)**2 * \
+                (d**2 + 2 * d * (r1 + r2) - 3 * (r1**2 + r2**2) + 6 * r1 * r2)
+        return vol / (4./3 * math.pi * min(r1, r2) ** 3)
```
Comment (jni) [inline skimage/feature/blob.py]: This function is kind of huge. How about:

```python
if ndim == 2:
    f = _compute_circle_overlap(d, r1, r2)
else:
    f = _compute_sphere_overlap(d, r1, r2)
return f
```

File: skimage/feature/blob.py [hunk 14]
```diff
@@ -91,15 +100,28 @@ def _prune_blobs(blobs_array, overlap):
...
+    if len(pairs) == 0:
+        return blobs_array
+    else:
+        for (i, j) in pairs:
+            blob1, blob2 = blobs_array[i], blobs_array[j]
+            if _blob_overlap(blob1, blob2) > overlap:
+                if blob1[-1] > blob2[-1]:
+                    blob2[-1] = -1
+                else:
+                    blob1[-1] = -1
```
Comment (jni) [inline skimage/feature/blob.py]: Can we just set to 0? -1 is ugly, and 0 does the job and is intuitive (making the blob "vanish").

File: skimage/feature/blob.py [hunk 15]
```diff
@@ -134,10 +156,12 @@ def blob_dog(image, min_sigma=1, max_sigma=50, sigma_ratio=1.6, threshold=2.0,
 
     Returns
     -------
-    A : (n, 3) ndarray
-        A 2d array with each row representing 3 values, ``(y,x,sigma)``
-        where ``(y,x)`` are coordinates of the blob and ``sigma`` is the
-        standard deviation of the Gaussian kernel which detected the blob.
+    A : (n, image.ndim + 1) ndarray
+        A 2d array with each row representing 3 values for a 2D image,
+        and 4 values for a 3D image: ``(r, c, sigma)`` or ``(f, r, c, sigma)``
```
Comment (jni) [inline skimage/feature/blob.py]: `f` -> `p`

File: skimage/feature/blob.py [hunk 16]
```diff
@@ -193,19 +216,24 @@ def blob_dog(image, min_sigma=1, max_sigma=50, sigma_ratio=1.6, threshold=2.0,
     # multiplying with standard deviation provides scale invariance
     dog_images = [(gaussian_images[i] - gaussian_images[i + 1])
                   * sigma_list[i] for i in range(k)]
-    image_cube = np.dstack(dog_images)
+
+    # Replace by image_cube = np.stack(hessian_images, axis=-1)
+    # When we upgrade minimal requirements to NumPy 1.10
+    sl = (slice(None),) * image.ndim + (np.newaxis,)
+    arrays = [np.asanyarray(arr) for arr in dog_images]
```
Comment (jni) [inline skimage/feature/blob.py]: LOL I hadn't seen this code before... `dog_images` definitely sounds like something other than what it is. =)

Comment (jni) [inline skimage/feature/blob.py]: @emmanuelle why are you putting the new axis at the end? This will be very inefficient...

File: skimage/feature/blob.py [hunk 17]
```diff
@@ -244,10 +272,12 @@ def blob_log(image, min_sigma=1, max_sigma=50, num_sigma=10, threshold=.2,
 
     Returns
     -------
-    A : (n, 3) ndarray
-        A 2d array with each row representing 3 values, ``(y,x,sigma)``
-        where ``(y,x)`` are coordinates of the blob and ``sigma`` is the
-        standard deviation of the Gaussian kernel which detected the blob.
+    A : (n, image.ndim + 1) ndarray
+        A 2d array with each row representing 3 values for a 2D image,
+        and 4 values for a 3D image: ``(r, c, sigma)`` or ``(f, r, c, sigma)``
```
Comment (jni) [inline skimage/feature/blob.py]: `f` -> `p`

File: skimage/feature/blob.py [hunk 18]
```diff
@@ -295,19 +323,24 @@ def blob_log(image, min_sigma=1, max_sigma=50, num_sigma=10, threshold=.2,
...
     # s**2 provides scale invariance
     gl_images = [-gaussian_laplace(image, s) * s ** 2 for s in sigma_list]
-    image_cube = np.dstack(gl_images)
+
+    # Replace by image_cube = np.stack(hessian_images, axis=-1)
+    # When we upgrade minimal requirements to NumPy 1.10
+    sl = (slice(None),) * image.ndim + (np.newaxis,)
+    arrays = [np.asanyarray(arr) for arr in gl_images]
+    extended_arrays = [arr[sl] for arr in arrays]
+    image_cube = np.concatenate(extended_arrays, axis=-1)
```
Comment (jni) [inline skimage/feature/blob.py]: As above -- concatenate on first axis, not last

**Comment by jni** [comment]
@emmanuelle Sorry for the delay! You can blame @NelleV for holding me up late at night with extended discussions about diversity. =P One more round of changes, please! They shouldn't be too hard. =)

File: skimage/feature/blob.py (hunk 10 above)
Comment (emmanuelle) [inline skimage/feature/blob.py]: Not sure I understand, you need to have an extra dimension for the computation of ``d = sqrt(np.sum((blob1[:-1] - blob2[:-1])**2))``

Comment (emmanuelle) [inline skimage/feature/blob.py]: Or the function should be refactored so that, in ``_prune_blobs``, one would pass only ``blob[:-1]``. What do you think?

File: skimage/feature/blob.py (hunk 18 above)
Comment (emmanuelle) [inline skimage/feature/blob.py]: I did it to stick to the original code, which was
```
gl_images = [-gaussian_laplace(image, s) * s ** 2 for s in sigma_list]
image_cube = np.dstack(gl_images)
```
so, concatenation along last axis. A very very small fraction of the time is spent on this operation anyway.

**Comment by emmanuelle** [comment]
No worries @jni, I would have loved to take part to such discussions :-)... but maybe @NelleV told you why I didn't come to Berkeley this year ;-). 

Thanks a lot for the comments, I implemented most of them, except for two points on which I'd like to have your opinion.

File: skimage/feature/blob.py (hunk 10 above)
Comment (jni) [inline skimage/feature/blob.py]: @emmanuelle no, one line down! "where `row, col` (or `(pln, row, col, sigma)`) are coordinates" -> "where `row, col` (or `(pln, row, col)`) are coordinates"

File: skimage/feature/blob.py (hunk 15 above)
Comment (jni) [inline skimage/feature/blob.py]: @emmanuelle this still needs updating!

File: skimage/feature/blob.py (hunk 18 above)
Comment (jni) [inline skimage/feature/blob.py]: Ugh. Very sad but I agree about backwards compatibility. =)

Comment (jni) [inline skimage/feature/blob.py]: Actually I just saw that these arrays are just used internally, so restacking wouldn't change the API... So I still think you should update. It's an antipattern to stack in this order in NumPy, so it's best if we avoid it in our own code! I think it simplifies the code a bit, too, to stack along first axis.

Plus you get to close the [issue](https://github.com/scikit-image/scikit-image/issues/2173) I just opened. =P

**Comment by jni** [comment]
@emmanuelle 

> but maybe @NelleV told you why I didn't come to Berkeley this year ;-).

Yes, she started with, "Oh, so you don't know why @emmanuelle couldn't come?" and had the **biggest** grin on her face. =D

> except for two points on which I'd like to have your opinion.

Now you have it! =P

**Comment by NelleV** [comment]
I still do =D

**Comment by rueberger** [comment]
It seems like this is ready to be merged?

File: skimage/feature/blob.py (hunk 15 above)
Comment (jni) [inline skimage/feature/blob.py]: @emmanuelle Quick ping! You need to change a couple of `f`s to `p`s!

File: skimage/feature/blob.py [hunk 19]
```diff
@@ -279,11 +359,9 @@ def blob_log(image, min_sigma=1, max_sigma=50, num_sigma=10, threshold=.2,
 
     Notes
     -----
-    The radius of each blob is approximately :math:`\sqrt{2}sigma`.
+    The radius of each blob is approximately :math:`\sqrt{2}sigma` for
+    a 2-D image and :math:`\sqrt{3}sigma` for a 3-D image.
```
Comment (jni) [inline skimage/feature/blob.py]: Should this be `\sqrt{3}\sigma`? If so it needs fixing in a few places.

File: skimage/feature/blob.py [hunk 20]
```diff
@@ -295,19 +373,24 @@ def blob_log(image, min_sigma=1, max_sigma=50, num_sigma=10, threshold=.2,
     # computing gaussian laplace
     # s**2 provides scale invariance
     gl_images = [-gaussian_laplace(image, s) * s ** 2 for s in sigma_list]
-    image_cube = np.dstack(gl_images)
+
+    # Replace by image_cube = np.stack(hessian_images, axis=-1)
+    # When we upgrade minimal requirements to NumPy 1.10
```
Comment (jni) [inline skimage/feature/blob.py]: @emmanuelle this has happened now so do you want to fix this now? =)

File: skimage/feature/blob.py [hunk 21]
```diff
@@ -321,7 +404,7 @@ def blob_doh(image, min_sigma=1, max_sigma=30, num_sigma=10, threshold=0.01,
 
     Parameters
     ----------
-    image : ndarray
+    image : 2D ndarray
```
Comment (jni) [inline skimage/feature/blob.py]: @emmanuelle the Hessian is now nD, so maybe you want to update this also? Would it be difficult? I wouldn't hold up this PR for it but it would be nice if you have some time! =)

**Comment by jni** [comment]
@rueberger almost! And LOL, I totally forgot about this PR, when @soupault linked to it I mistakenly thought it was an issue. GitHub should visually differentiate the two a bit more. =P

**Comment by rueberger** [comment]
bump :) this issue has come up again for me and it would be really nice to have support for 3D blob_log

**Comment by jni** [comment]
@rueberger I might have time next week, but feel free to grab this branch, make the requested changes, and submit your own PR! =)

**Comment by soupault** [comment]
This should definitely go into 0.14. (Just saying :) )

**Comment by jni** [comment]
Closing in favor of #2854
